rfd/0134-cloud-first-deployment.md

```diff
@@ -0,0 +1,182 @@
+---
+authors: Roman Tkachenko (roman@goteleport.com),
+         Zac Bergquist (zac@goteleport.com),
+         Jim Bishopp (jim@goteleport.com),
+         RJ (rjones@goteleport.com),
+         Sasha Klizhentas (sasha@goteleport.com)
+state: draft
+---
+
+# RFD 134 - Cloud-first deployment
+
+## Required approvers
+
+- Engineering: @russjones, @jimbishopp, @zmb3
+- Product: @klizhentas, @xinding33
+
+## What
+
+This RFD describes the Cloud-first deployment approach, as well as the release
+process changes required to support it.
+
+## Why
+
+Our current feature delivery process is not compatible with the company's shift
+to "SaaSify" Teleport. The new (or rather, extended) Cloud-focused process aims
+to address shortcomings like:
+
+- Long release cycles where marquee new features get delivered to Cloud users
+  (roughly) once a quarter and are tied to major self-hosted releases.
+- Slow and monolithic release pipeline that always builds all artifacts and
+  does not allow for very fast Cloud rollouts.
+- Engineers being disconnected from Cloud deployments and unaware of when their
+  changes are getting rolled out.
+
+## Self-hosted release process (current)
+
+Before diving into the new process, let's set the stage with a high-level overview
+of the current release process.
+
+Current release versioning scheme follows semver, with quarterly major releases
+and monthly minor releases.
+
+Major releases are getting rolled onto Cloud quarterly, with a month's delay.
+Major features are reserved for major releases, which leads to Cloud customers
+getting major features with a month's delay.
+
+The release process is triggered on a Github tag creation and builds the full
+suite of artifacts that takes 2-3 hours so it's unrealistic to do multiple
+deploys a day.
+
+Cloud upgrade happens once a week to a release published the previous week, and
+engineers oftentimes don't have a clear idea of when their change is being
+deployed.
+
+## Cloud release process
+
+In the Cloud-first world, release that's running on Cloud does not necessarily
+have a corresponding self-hosted release tag. For this reason, we can't reuse
+the current versioning scheme and apply it to both Cloud and self-hosted.
+
+Instead, we want to get to a place where Cloud and self-hosted releases are
+decoupled, with Cloud releases being regularly cut from the main development
+branch and self-hosted releases cut from their respective release branches.
+
+Although the concept of semantic versioning loses most of its meaning with the
+continuous release train, we do need to keep it for Cloud releases to be able
+to use automatic upgrades system we built.
+
+With this in mind, the Cloud version will consist of:
+
+- Major: "development" version number on master branch (latest stable release + 1).
+- Minor/patch: always 0.
+- Prerelease: `-cloud.X` with a perpetually increasing "build" number which will
```

Comment:
> might as well drop semver altogether

Our auto upgrades machinery requires valid semver, which is why we're using 14.0.0 even though the minor/patch versions don't convey anything.

Reply:
A reason for the suggestion is that semver makes little sense for CD, it's usually just a bother, plus I imagine we could patch the tooling to do what we wanted. That said we can surely live the zeroes. The argument for the commit hash is a bit stronger, I think, but up to y'all.

Reply:
I don't have strong feelings about the commit hash - it has the nice property of being easily tied to a commit, but it does make it harder to know at a glance if a tenant running `cloud.9da628ffcb` is newer or older than one running `cloud.27941f402`.

Reply:
If we use hash, then we lose ability to compare and our auto-upgrades stop working. I think counter is fine for now, you can always lookup exact commit by tag.

Reply:
Gotcha, makes sense. Feel free to resolve the thread.

Reply:
Adding to this can of worms to suggest (perhaps, terribly) a timestamp rather than a counter, e.g. `v14.0.0-cloud.2023-07-12T04:05:48Z`

Pro:
- still easy to identify newer vs. older revisions of Teleport
- still sorts in temporal order of releases
- makes it possible to speak about compatibility windows in terms of _X days_ or _Y weeks_, which may make more sense to humans than _X releases_, especially if releases are irregular.

Con:
- I have no science to back up the idea that talking in wall-clock time is any more useful than talking in releases
- It's very verbose
- It's just a bit odd

Reply:
Cloud has been tagging infra releases like `prod-aabbccd-20230712T123633`, we could flip it around and use `v14.0.0-cloud.20230712T123633.aabbccd` to have both time-based ordering that matches the lexicographic one and a reference for the exact commit that's running.

Reply:
+1 for commit+timestamp - it has the bulk of what we care about without the need for counter bookkeeping. (Extra points if on rollbacks we don't "update" the timestamp.)

Reply:
Sounds good, I agree this looks like better versioning scheme. Updated.